File: Zhengyingying/src/com/zyy/week6/BinaryTreeZigzagLevelOrderTraversal.java
package com.zyy.week6;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class BinaryTreeZigzagLevelOrderTraversal {
	// 宽度优先搜索BFS
	public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
		List<List<Integer>> result = new ArrayList<List<Integer>>();
		if (root == null)
			return result ;
		Queue<TreeNode> queue = new LinkedList<TreeNode>();
		//保证顺序
		LinkedList<Integer> valList = new LinkedList<Integer>();
		int levelNumber = 1;
		boolean flag = true;
		queue.add(root);
		TreeNode node;
		while (!queue.isEmpty()) {
			node = queue.poll();
			if (flag)
				valList.add(node.val);
			else
				valList.addFirst(node.val);

			if (node.left != null) {
				queue.add(node.left);
			}
			if (node.right != null) {
				queue.add(node.right);
			} 

			--levelNumber;
			if (levelNumber == 0) {
				result.add(valList);
				//这一层，有几个需要处理的
				levelNumber = queue.size();
				//不能使用clear方法：这里都是指针，只是clear还是指的原来的list，如果要新的，就得重新创建
				valList= new LinkedList<Integer>();
				flag = !flag;
			}
		}
		return result;
	}

}

File: TangHongyin/binary_tree_zigzag_level_order_traversal.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):

    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = [[]]
        v_result = [[]]
        flag = True
        layer = 0

        if root is not None:
            result[layer].append(root)
            v_result[layer].append(root.val)
        else:
            return []

        while layer < len(result):
            queue = []
            v_queue = []
            flag = not flag
            # from right to left
            if flag is not True:
                for i in range(len(result[layer]) - 1, -1, -1):
                    if result[layer][i].right is not None:
                        queue.insert(0, result[layer][i].right)
                        v_queue.append(result[layer][i].right.val)
                    if result[layer][i].left is not None:
                        queue.insert(0, result[layer][i].left)
                        v_queue.append(result[layer][i].left.val)
                    # if result[layer][i].left is not None:
                    #     queue.append(result[layer][i].left)
                    # if result[layer][i].right is not None:
                    #     queue.append(result[layer][i].right)
            else:
                for i in range(len(result[layer])):
                    if result[layer][i].left is not None:
                        queue.append(result[layer][i].left)
                        v_queue.append(result[layer][i].left.val)
                    if result[layer][i].right is not None:
                        queue.append(result[layer][i].right)
                        v_queue.append(result[layer][i].right.val)
            if len(queue) > 0:
                result.append(queue)
                v_result.append(v_queue)
            layer += 1
        return v_result
